card/mod/core/set/all/fetch_helper.rb

```diff
@@ -72,17 +74,22 @@ def retrievable_mark_type_and_value mark
     end
   end
 
-  def retrieve_from_db? card, mark_type, mark_key, look_in_trash
-    return false if card&.real?                                      # real card found in cache
-#    return false if mark_type == :key && mark_key.to_name.relative?  # no relative names in db
-    card.nil? || (look_in_trash && !card.trash)                      # card not found in cache (or trash lookup)
+  def retrieve_from_db card, mark_type, mark_key, opts
+    look_in_trash = opts[:look_in_trash]
+    return [card, false] unless retrieve_from_db? card, look_in_trash
+    card = retrieve_from_db! mark_type, mark_key, look_in_trash
+    needs_caching = !card.nil? && !card.trash
+    [card, needs_caching]
   end
 
-  def retrieve_from_db mark_type, mark_key, look_in_trash=false
+  def retrieve_from_db? card, look_in_trash
+    card.nil? || (look_in_trash && !card.trash)
+  end
+
+  def retrieve_from_db! mark_type, mark_key, look_in_trash=false
```

Comment:
why the bang?

Reply:
because there's already a `retrieve_from_db` (bad answer).  will rename other method and drop this bang.